core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala

```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
   }
 
-  def waitingForRoutee(work: Option[(Work, ActorRef)]): Receive = whileWaiting(Starting) orElse {
-    case RouteeReceived(r) ⇒
-      routee = r
-      context watch r
-      context become waitingForWork
-      work match {
-        case Some((workMsg, workSender)) ⇒ self.tell(workMsg, workSender)
-        case None                        ⇒ askMoreWork()
-      }
+  def working(outstanding: Outstanding): Receive = routeeResponse(outstanding, becomeUnregistering) orElse {
+    case qs: QueryStatus     ⇒ qs reply Working
+    case Hold(period)        ⇒ delayBeforeNextWork = Some(period)
 
-    //this only happens on restarting with new routee because the old one died.
-    case w: Work ⇒
-      if (work.isDefined) {
-        sender() ! Rejected(w, "Busy")
-      } else {
-        context become waitingForRoutee(Some((w, sender)))
-      }
+    //we are done with this Work, ask for more and wait for it
+    case WorkFinished        ⇒ askMoreWork()
+    case w: Work             ⇒ sender() ! Rejected(w, "Busy")
 
-    case Worker.Retire ⇒
-      work.foreach {
-        case (workMsg, workSender) ⇒ workSender ! Rejected(workMsg, "retiring") //this happens only when a Retire is sent during a restart with new routee
-      }
-      finish()
+    //if there is no work left, or if the Queue dies, the Actor must wait for the Work to finish before terminating
+    case Terminated(`queue`) ⇒ context become waitingToTerminate(outstanding)
+    case NoWorkLeft          ⇒ context become waitingToTerminate(outstanding)
 
+    //This is a fun state.  The Worker is told to stop, but needs to both wait for Unregister and for Work to complete
+    case Worker.Retire ⇒
+      queue ! Unregister
+      context become unregisteringOutstanding(outstanding)
   }
 
-  val waitingForWork: Receive =
-    whileWaiting(Idle) orElse {
+  //This state waits for Work to complete, and then stops the Actor
+  def waitingToTerminate(outstanding: Outstanding): Receive = routeeResponse(outstanding, finish) orElse {
+    case qs: QueryStatus     ⇒ qs reply WaitingToTerminate
 
-      case work: Work ⇒ sendWorkToDelegatee(work, 0)
+    //ignore these, since all we care about is the Work completing one way or another
+    case Hold(x)             ⇒
+    case Retire              ⇒
+    case Terminated(`queue`) ⇒
+    case NoWorkLeft          ⇒
 
-      case Terminated(r) if r == routee ⇒
-        context become waitingForRoutee(None)
-        retrieveRoutee()
+    case w: Work             ⇒ sender() ! Rejected(w, "Retiring") //safety first
 
-      case Worker.Retire ⇒
-        queue ! Unregister(self)
-        context become retiring(None)
-    }
+    case WorkFinished        ⇒ finish() //work is done, terminate
+  }
 
-  def whileWaiting(currentStatus: WorkerStatus): Receive = {
-    case NoWorkLeft          ⇒ finish()
+  //in this state, we have told the Queue to Unregister this Worker, so we are waiting for an acknowledgement
+  def unregistering: Receive = {
```

Comment:
What if we changed these names to both `unregisteringIdle` and `unregisteringBusy`, in the spirit of brevity